### PR TITLE
Add extract declarations concept to CumulusCI

### PR DESCRIPTION
**Pull request by prescod**
Add extract declarations concept to CumulusCI.

Documentation will be added later when we unveil to the public.

### CODE REVIEW

**Review by davidmreed — COMMENTED** [review]
This looks great. I love the expansion logic. A few questions below but not major concerns about the feature.

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from .extract_yml import ExtractDeclaration
+
+_DEFAULT_DECLARATIONS = [
+    ExtractDeclaration(
+        sf_object="Account", where="Name != 'Sample Account for Entitlements'"
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:5]: This Sample Data is populated when hasSampleData is False:

https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000094mBUIAY/view

https://trailblazer.salesforce.com/issues_view?id=a1p4V000001qQlyQAE&title=entitlements-sample-data-being-created-in-spring-21-scratch-orgs-when-hassampledata-is-set-to-false

We have accidentally extracted it:

https://github.com/SalesforceFoundation/NPSP/blob/b65301d7609303fecc72ff8a6dd96c6d55c8cbe7/datasets/dev_org/test_data.sql#L14

I have spent many hours working around this bug:

https://github.com/SalesforceFoundation/NPSP/blob/2a7b1f08914fcb61198b0d42d22019f6a645f16e/robot/Cumulus/perftests/BDI_Tests.robot#L80

😢

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        where="Name NOT IN ('First Response to Customer', 'Escalate Case', 'Close Case')",
+    ),
+    ExtractDeclaration(
+        sf_object="WorkBadgeDefinition",
+        where="Name NOT IN ('Thanks', 'You\\'re a RockStar!', 'Team Player',"
+        "'All About Execution', 'Teacher', 'Top Performer', 'Hot Lead',"
+        "'Key Win', 'Customer Hero', 'Competition Crusher',"
+        "'Deal Maker', 'Gold Star')",
+    ),
+    ExtractDeclaration(
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:37]: For example, here are the relevant bits of EmailTemplate SObject:

```python
{
    "activateable": False,
    "compactLayoutable": True,
    "createable": True,
    "deepCloneable": False,
    "deletable": True,
    "deprecatedAndHidden": False,
    "feedEnabled": False,
    "hasSubtypes": False,
    "isInterface": False,
    "isSubtype": False,
    "layoutable": True,
    "mergeable": False,
    "mruEnabled": True,
    "name": "EmailTemplate",
    "queryable": True,
    "replicateable": True,
    "retrieveable": True,
    "searchLayoutable": False,
    "searchable": True,
    "triggerable": False,
    "undeletable": False,
    "updateable": True,
}
```

File: cumulusci/tasks/bulkdata/extract_dataset_utils/synthesize_extract_declarations.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    )
+    return simple_declarations
+
+
+def _merge_group_declarations_with_simple_declarations(
+    simple_declarations: T.Iterable[ExtractDeclaration],
+    group_declarations: T.Iterable[ExtractDeclaration],
+    schema: Schema,
+) -> T.List[ExtractDeclaration]:
+    """Expand group declarations to simple declarations and merge
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/synthesize_extract_declarations.py:83]: Actually deciding what we want to do in this case is an open question for me at this point. "Union" semantics or "Override" semantics. Trying to decide which is more useful/understandable/composable.

Doesn't matter for Metecho use-case but when these are hand-authored...

File: cumulusci/tasks/bulkdata/extract_dataset_utils/synthesize_extract_declarations.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+        def matches_obj(obj):
+            return obj.custom
+
+    elif decl.group_type == SFObjectGroupTypes.all:
+
+        def matches_obj(obj):
+            return True
+
+    else:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/synthesize_extract_declarations.py:120]: It's Coverage that I'm trying to help.

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+from .extract_yml import ExtractDeclaration
+
+_DEFAULT_DECLARATIONS = [
+    ExtractDeclaration(
+        sf_object="Account", where="Name != 'Sample Account for Entitlements'"
+    ),
+    ExtractDeclaration(sf_object="BusinessHours", where="Name != 'Default'"),
+    ExtractDeclaration(sf_object="ContentWorkspace", where="Name != 'Asset Library'"),
+    ExtractDeclaration(sf_object="Entitlement", where="Name != 'Sample Entitlement'"),
+    ExtractDeclaration(
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:10]: Yep.

```
>>> sf.FieldServiceMobileSettings.describe()["createable"]
True
```

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:37]:
```
>>> [getattr(sf, obj).describe()["createable"] for obj in ("EmailTemplate", "Folder", "Weblink", "MilestoneType")]
[True, True, True, True]
```

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+from .extract_yml import ExtractDeclaration
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:1]: I propose a new WI that evaluates the detailed questions of dynamically merging:

* built-in declarations
* generic declarations (OBJECT(STANDARD))
* specific declarations

Metecho doesn't need this level of sophistication so I think we can punt it a bit.

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:37]: Wild. Thanks for confirming.

File: cumulusci/tasks/bulkdata/extract_dataset_utils/synthesize_extract_declarations.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/synthesize_extract_declarations.py:93]: I don't think I have strong opinions in any direction, except that union semantics is likely more complicated to implement. Override seems generally reasonable. Also reasonable to say that dynamic declarations just don't expand across objects that are already in the definition.

File: cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/extract_dataset_utils/hardcoded_default_declarations.py:5]: That is... awful. Maybe I will harangue someone about this. Withdrawing my comment here.

**Review by davidmreed — APPROVED** [review]

**Pull request merged**